packages/contracts-bedrock/src/L1/OPStackManager.sol

```diff
@@ -0,0 +1,401 @@
+    bytes32 public latestRelease;
+
+    /// @notice Maps a release version to a contract name to it's implementation data.
+    mapping(bytes32 => mapping(bytes32 => Implementation)) public implementations;
+
+    /// @notice Maps an L2 Chain ID to the SystemConfig for that chain.
+    /// Most information for a chain can be found from it's the SystemConfig, with ProtocolVersions
+    /// being the exception. ProtocolVersions is a singleton for all chains deployed by this
+    /// contract, so `referenceProtocolVersions` is used for all chains.
+    /// Additionally, the ProxyAdmin can be discovered offchain, but cannot be discovered onchain
+    /// due to implementation details of the ProxyAdmin contract.
+    mapping(uint256 => SystemConfig) public systemConfigs;
+
+    /// @notice Maps an L2 Chain ID to the release version for that chain.
+    /// Release strings follow semver and are named with the format `op-contracts/vX.Y.Z`.
+    mapping(uint256 => bytes32) public releases;
+
+    constructor(
+        uint64 referenceChainId,
+        SystemConfig _referenceSystemConfig,
+        ProtocolVersions _referenceProtocolVersions,
+        AddressManager _referenceAddressManager,
+        address _referenceProxyAdminOwner
+    ) {
+        _disableInitializers();
+
+        referenceSystemConfig = _referenceSystemConfig;
+        referenceProtocolVersions = _referenceProtocolVersions;
+
+        address referencePortal = referenceSystemConfig.optimismPortal();
+        referenceSuperchainConfig = OptimismPortal(payable(referencePortal)).superchainConfig();
+        referenceAddressManager = _referenceAddressManager;
+        referenceProxyAdminOwner = _referenceProxyAdminOwner;
+
+        register(referenceChainId, latestRelease, referenceSystemConfig);
+    }
+
+    function initialize(address _owner) external initializer {
+        __Ownable_init();
+        transferOwnership(_owner);
+    }
+
+    /// @notice Called by the OP Stack Manager owner to release a set of implementation contracts.
+    function release(bytes32 version, bool isLatest, ImplementationSetter[] calldata impls) public onlyOwner {
+        for (uint256 i = 0; i < impls.length; i++) {
+            ImplementationSetter calldata implSetter = impls[i];
+            Implementation storage impl = implementations[version][implSetter.name];
+            require(impl.logic == address(0), "OpStackManager: Implementation already exists");
+
+            impl.initializer = implSetter.info.initializer;
+            impl.logic = implSetter.info.logic;
+        }
+
+        if (isLatest) {
+            latestRelease = version;
+        }
+    }
+
+    /// @notice Used to make this contract aware of chains that were deployed without this factory.
+    /// This method is permissionless, and anyone can register any chain ID.
+    function register(uint64 l2ChainId, bytes32 release_, SystemConfig sc) public {
```

Comment:
Long term we were thinking of having a way to derive a chain id given a chain config, to do this a hash of the chain config could be used, i think chain ids needs to by uint256, also the interop spec defines them as uint256

Reply:
I went with uint64 so we can do this
https://github.com/ethereum-optimism/optimism/blob/50a0125bad781f915ac2008738763230718920ec/packages/contracts-bedrock/src/L1/OPStackManager.sol#L301-L304

We could still do a hash of the chain config and truncate to uint64 and are unlikely to have a collision, but that's outside the scope of this PR. Ref https://github.com/ethereum-optimism/specs/issues/94 where this is being tracked. 

Either way, this contract is upgradeable so we can always start with uint64 to keep the existing batch inbox format until https://github.com/ethereum-optimism/specs/issues/94 is finalized with e.g. a chain config approach.